### PR TITLE
Add System.Security.Cryptography.Hashing.Algorithms Source and Test

### PR DESCRIPTION
**Pull request by bartonjs**
Resolves https://github.com/dotnet/corefx/issues/836.

### CODE REVIEW

File: src/Common/src/Interop/Windows/BCrypt/Cng.cs [hunk 1]
```diff
@@ -0,0 +1,394 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using global::System;
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/BCrypt/Cng.cs]: Can we clean up these "global::"s?  It's ok to do so in a separate commit.

File: src/Common/src/Interop/Windows/BCrypt/Cng.cs [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+            public const string Hash = "HASH";                  // BCRYPT_KDF_HASH
+            public const string Hmac = "HMAC";                  // BCRYPT_KDF_HMAC
+            public const string Tls = "TLS_PRF";                // BCRYPT_KDF_TLS_PRF
+        }
+    }
+
+    //
+    // Interop layer around Windows CNG api.
+    //
+    internal static partial class Cng
```
Comment (stephentoub) [inline src/Common/src/Interop/Windows/BCrypt/Cng.cs]: In general there's probably some clean up to do in these interop files.  I'd suggest just opening an issue to deal with it separately.

File: src/System.Security.Cryptography.Hashing.Algorithms/tests/MD5Tests.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        {
+            return MD5.Create();
+        }
+
+        // Test cases are defined in RFC 1321, section A.5
+
+        [Fact]
+        public void MD5_Rfc1321_1()
+        {
+            Verify("", "d41d8cd98f00b204e9800998ecf8427e");
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Hashing.Algorithms/tests/MD5Tests.cs]: As I mentioned offline, we could consider using [Theory]/[InlineData] for these, but I understand we're currently limited with the internal test harness.

File: src/System.Security.Cryptography.Hashing.Algorithms/tests/System.Security.Cryptography.Hashing.Algorithms.Tests.csproj [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    <RootNamespace>System.Security.Cryptography.Hashing.Algorithms.Tests</RootNamespace>
+  </PropertyGroup>
+  <!-- Default configurations to help VS understand the configurations -->
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+  </PropertyGroup>
+  <ItemGroup>
+    <!-- Compile tests against the contract, but copy our local-built implementation for testing -->
+    <ProjectReference Include="..\src\System.Security.Cryptography.Hashing.Algorithms.csproj">
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Hashing.Algorithms/tests/System.Security.Cryptography.Hashing.Algorithms.Tests.csproj]: This is the mechanism used for partial facades, which this isn't, right?  We should be able to use a normal P2P reference without the rest of the below gook.

**Comment by stephentoub** [comment]
Other than the structure of the P2P reference in the tests project, which I think can be updated to match how P2P references are done in other non-partial facades, LGTM.

File: src/Common/src/Interop/Windows/BCrypt/Cng.cs (hunk 1 above)
Comment (bartonjs) [inline src/Common/src/Interop/Windows/BCrypt/Cng.cs]: Fixed, squashed commit.

File: src/Common/src/Interop/Windows/BCrypt/Cng.cs (hunk 2 above)
Comment (bartonjs) [inline src/Common/src/Interop/Windows/BCrypt/Cng.cs]: Created https://github.com/dotnet/corefx/issues/1739 to track the cleanup.

File: src/System.Security.Cryptography.Hashing.Algorithms/tests/System.Security.Cryptography.Hashing.Algorithms.Tests.csproj (hunk 4 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.Hashing.Algorithms/tests/System.Security.Cryptography.Hashing.Algorithms.Tests.csproj]: Must have picked a bad example at random.  Changed the reference style to match System.Xml.XmlDocument and System.Diagnostics.FileVersionInfo. (Squashed commit)

**Comment by stephentoub** [comment]
Thanks. LGTM.

File: src/System.Security.Cryptography.Hashing.Algorithms/src/System.Security.Cryptography.Hashing.Algorithms.csproj [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    <CLSCompliant>false</CLSCompliant>
+  </PropertyGroup>
+  <!-- Help VS understand available configurations -->
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Linux_Debug|AnyCPU' " />
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Linux_Release|AnyCPU' " />
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'OSX_Debug|AnyCPU' " />
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'OSX_Release|AnyCPU' " />
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Windows_Debug|AnyCPU' " />
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Windows_Release|AnyCPU' " />
+  <!-- References are resolved from packages.config -->
```
Comment (akoeplinger) [inline src/System.Security.Cryptography.Hashing.Algorithms/src/System.Security.Cryptography.Hashing.Algorithms.csproj]: ~~packages.config~~ project.json. Should probably also move it to the respective ItemGroup below.

File: src/System.Security.Cryptography.Hashing.Algorithms/tests/System.Security.Cryptography.Hashing.Algorithms.Tests.csproj [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    <Compile Include="HmacTests.cs" />
+    <Compile Include="MD5Tests.cs" />
+    <Compile Include="Rfc2202HmacTests.cs" />
+    <Compile Include="Rfc4231HmacTests.cs" />
+    <Compile Include="Sha1Tests.cs" />
+    <Compile Include="Sha256Tests.cs" />
+    <Compile Include="Sha384Tests.cs" />
+    <Compile Include="Sha512Tests.cs" />
+  </ItemGroup>
+  <!-- References are resolved from packages.config -->
```
Comment (akoeplinger) [inline src/System.Security.Cryptography.Hashing.Algorithms/tests/System.Security.Cryptography.Hashing.Algorithms.Tests.csproj]: ~~packages.config~~ project.json

**Comment by stephentoub** [comment]
Jeremy, ideally we'll get the Linux CI machines configured correctly for this asap, but if you don't want to block and would prefer to check in, you could temporarily suppress the Linux tests by adding a line like this to the test project:
https://github.com/dotnet/corefx/blob/master/src/Microsoft.Win32.Registry/tests/Microsoft.Win32.Registry.Tests.csproj#L11

File: src/System.Security.Cryptography.Hashing.Algorithms/src/System.Security.Cryptography.Hashing.Algorithms.csproj (hunk 5 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.Hashing.Algorithms/src/System.Security.Cryptography.Hashing.Algorithms.csproj]: Did a check across the repo and it's hit-or-miss on whether the comment was there at all (the location here is where VS would normally write in the references), so I went ahead and just removed the lines in both files.
